Use Future's then() API to fix RPC profiling

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38352 Use Future's then() API to fix RPC profiling**

Fixes the RPC profiling by using the `then()` API added in https://github.com/pytorch/pytorch/pull/37311. Instead of adding a regular callback, we return a new future that completes when the profiling callback is finished. This is transparent to the user as the future still completes with the value of the original future (i.e. the RPC's return value)

To make this work for RRef, we add a `_set_profiling_future` to set the profiling future, and `_get_profiling_future` to retrieve this future and wait on it in the tests.

Re-enabled profiling tests and stress tested them 1000 times to verify the fix

Differential Revision: [D21506940](https://our.internmc.facebook.com/intern/diff/D21506940/)

rohan-varma committed:
Use Future's then() API to fix RPC profiling

Fixes the RPC profiling by using the `then()` API added in https://github.com/pytorch/pytorch/pull/37311. Instead of adding a regular callback, we return a new future that completes when the profiling callback is finished. This is transparent to the user as the future still completes with the value of the original future (i.e. the RPC's return value)

To make this work for RRef, we add a `_set_profiling_future` to set the profiling future, and `_get_profiling_future` to retrieve this future and wait on it in the tests.

Re-enabled profiling tests and stress tested them 1000 times to verify the fix

Differential Revision: [D21506940](https://our.internmc.facebook.com/intern/diff/D21506940/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Use Future's then() API to fix RPC profiling"


Fixes the RPC profiling by using the `then()` API added in https://github.com/pytorch/pytorch/pull/37311. Instead of adding a regular callback, we return a new future that completes when the profiling callback is finished. This is transparent to the user as the future still completes with the value of the original future (i.e. the RPC's return value)

To make this work for RRef, we add a `_set_profiling_future` to set the profiling future, and `_get_profiling_future` to retrieve this future and wait on it in the tests.

Re-enabled profiling tests and stress tested them 1000 times to verify the fix

Differential Revision: [D21506940](https://our.internmc.facebook.com/intern/diff/D21506940/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Use Future's then() API to fix RPC profiling"


Fixes the RPC profiling by using the `then()` API added in https://github.com/pytorch/pytorch/pull/37311. Instead of adding a regular callback, we return a new future that completes when the profiling callback is finished. This is transparent to the user as the future still completes with the value of the original future (i.e. the RPC's return value)

To make this work for RRef, we add a `_set_profiling_future` to set the profiling future, and `_get_profiling_future` to retrieve this future and wait on it in the tests.

Re-enabled profiling tests and stress tested them 1000 times to verify the fix

Differential Revision: [D21506940](https://our.internmc.facebook.com/intern/diff/D21506940/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Use Future's then() API to fix RPC profiling"


Fixes the RPC profiling by using the `then()` API added in https://github.com/pytorch/pytorch/pull/37311. Instead of adding a regular callback, we return a new future that completes when the profiling callback is finished. This is transparent to the user as the future still completes with the value of the original future (i.e. the RPC's return value)

To make this work for RRef, we add a `_set_profiling_future` to set the profiling future, and `_get_profiling_future` to retrieve this future and wait on it in the tests.

Re-enabled profiling tests and stress tested them 1000 times to verify the fix

Differential Revision: [D21506940](https://our.internmc.facebook.com/intern/diff/D21506940/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Use Future's then() API to fix RPC profiling"


Fixes the RPC profiling by using the `then()` API added in https://github.com/pytorch/pytorch/pull/37311. Instead of adding a regular callback, we return a new future that completes when the profiling callback is finished. This is transparent to the user as the future still completes with the value of the original future (i.e. the RPC's return value)

To make this work for RRef, we add a `_set_profiling_future` to set the profiling future, and `_get_profiling_future` to retrieve this future and wait on it in the tests.

Re-enabled profiling tests and stress tested them 1000 times to verify the fix

Differential Revision: [D21506940](https://our.internmc.facebook.com/intern/diff/D21506940/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -68,7 +66,11 @@ void _call_end_callbacks_on_fut(
         auto& rec = getRecordFunctionFromTensor(handle);
         rec._end();
-      });
+        return fut->constValue();
```
Let's add a comment to explain why we need to forward the return value here. IIUC, it is because `_invoke_rpc` actually returns this future?

rohan-varma replied:
Yeah, we get the future which has the actual result from `_invoke_rpc`. Then, if profiling we run this, and we want to ensure that the correct value of the future from `_invoke_rpc` is propagated. Will add this to the comment

mrshenli commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -84,12 +86,14 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
-          _call_end_callbacks_on_fut(tensor, fut);
+          auto profiledFut = _call_end_callbacks_on_fut(tensor, fut);
+          // return future that completes when profiling callbacks have run.
+          jit::push(stack, profiledFut);
```
IIUC, `push` takes rvalue reference. If so, shall we use `std::move()` on the `profilerFut`?

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -337,6 +337,29 @@ PyObject* rpc_init(PyObject* /* unused */) {
                   on the remote node. This is for internal use cases such as profiling
                   only.
               )")
+          .def(
+              "_get_profiling_future",
+              [](const PyRRef& self) {
+                return std::make_shared<jit::PythonFutureWrapper>(
+                    self.getProfilingFuture());
+              },
+              py::call_guard<py::gil_scoped_release>(),
+              R"(
+                  Returns future that completes when the profiling event corresponding
+                  to the creation of this RRef on the remote node has been recorded.
+              )")
+          .def(
+              "_set_profiling_future",
+              [](PyRRef& self,
+                 const std::shared_ptr<jit::PythonFutureWrapper>&
+                     wrappedFuture) {
+                self.setProfilingFuture(wrappedFuture->fut);
+              },
+              py::call_guard<py::gil_scoped_release>(),
```
These two seems are very light functions, should I hold GIL instead to avoid ctx switch?

mrshenli reviewed (approved):
Thanks for fixing! I added some minor comments.

mrshenli reviewed (changes requested):
Lint failure is real.

```
  {
    path: 'torch/testing/_internal/distributed/rpc/jit/rpc_test.py',
    start_line: 2,
    end_line: 2,
    start_column: 1,
    end_column: 1,
    annotation_level: 'failure',
    message: "[F401] 'unittest' imported but unused"
  }
```

rohan-varma committed:
Update on "Use Future's then() API to fix RPC profiling"


Fixes the RPC profiling by using the `then()` API added in https://github.com/pytorch/pytorch/pull/37311. Instead of adding a regular callback, we return a new future that completes when the profiling callback is finished. This is transparent to the user as the future still completes with the value of the original future (i.e. the RPC's return value)

To make this work for RRef, we add a `_set_profiling_future` to set the profiling future, and `_get_profiling_future` to retrieve this future and wait on it in the tests.

Re-enabled profiling tests and stress tested them 1000 times to verify the fix

Differential Revision: [D21506940](https://our.internmc.facebook.com/intern/diff/D21506940/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Let's land!
